In [2]:
import pandas as pd
import numpy as np


In [19]:
# data
full_games_data
pred_sum
# data
# week_games_sum
points
week_games_sum
data[data['screen_name']=='Market'].groupby(['game_id'])['game_id'].count()
full_games_data

game_id  season game_type  week     gameday weekday gametime  \
0     1999_01_MIN_ATL    1999       REG     1  1999-09-12  Sunday      NaN   
1      1999_01_KC_CHI    1999       REG     1  1999-09-12  Sunday      NaN   
2     1999_01_PIT_CLE    1999       REG     1  1999-09-12  Sunday      NaN   
3      1999_01_OAK_GB    1999       REG     1  1999-09-12  Sunday      NaN   
4     1999_01_BUF_IND    1999       REG     1  1999-09-12  Sunday      NaN   
...               ...     ...       ...   ...         ...     ...      ...   
6119   2021_18_CAR_TB    2021       REG    18  2022-01-09  Sunday    13:00   
6120  2021_18_SEA_ARI    2021       REG    18  2022-01-09  Sunday    16:25   
6121   2021_18_KC_DEN    2021       REG    18  2022-01-09  Sunday    16:25   
6122    2021_18_SF_LA    2021       REG    18  2022-01-09  Sunday    16:25   
6123   2021_18_LAC_LV    2021       REG    18  2022-01-09  Sunday    16:25   

     away_team  away_score home_team  home_score location  result  total  \
0          MIN        17.0       ATL        14.0     Home    -3.0   31.0   
1           KC        17.0       CHI        20.0     Home     3.0   37.0   
2          PIT        43.0       CLE         0.0     Home   -43.0   43.0   
3          OAK        24.0        GB        28.0     Home     4.0   52.0   
4          BUF        14.0       IND        31.0     Home    17.0   45.0   
...        ...         ...       ...         ...      ...     ...    ...   
6119       CAR         NaN        TB         NaN     Home     NaN    NaN   
6120       SEA         NaN       ARI         NaN     Home     NaN    NaN   
6121        KC         NaN       DEN         NaN     Home     NaN    NaN   
6122        SF         NaN        LA         NaN     Home     NaN    NaN   
6123       LAC         NaN        LV         NaN     Home     NaN    NaN   

      overtime  old_game_id     gsis                         nfl_detail_id  \
0          0.0   1999091210    598.0                                   NaN   
1          0.0   1999091206    597.0                                   NaN   
2          0.0   1999091213    604.0                                   NaN   
3          0.0   1999091208    602.0                                   NaN   
4          0.0   1999091202    591.0                                   NaN   
...        ...          ...      ...                                   ...   
6119       NaN   2022010911  58770.0  10160000-0587-704e-6295-0c29cd0a1891   
6120       NaN   2022010912  58771.0  10160000-0587-7113-a978-455fec3cda14   
6121       NaN   2022010913  58772.0  10160000-0587-7224-5e5e-c71409627e21   
6122       NaN   2022010914  58773.0  10160000-0587-738e-2b05-95daf03cb563   
6123       NaN   2022010915  58774.0  10160000-0587-7450-0923-b0f39c686c28   

               pfr  pff       espn  away_rest  home_rest  away_moneyline  \
0     199909120atl  NaN  190912001          7          7             NaN   
1     199909120chi  NaN  190912003          7          7             NaN   
2     199909120cle  NaN  190912005          7          7             NaN   
3     199909120gnb  NaN  190912009          7          7             NaN   
4     199909120clt  NaN  190912011          7          7             NaN   
...            ...  ...        ...        ...        ...             ...   
6119  202201090tam  NaN  401326596          7          7             NaN   
6120  202201090crd  NaN  401326597          7          7             NaN   
6121  202201090den  NaN  401326598          7          7             NaN   
6122  202201090ram  NaN  401326599          7          7             NaN   
6123  202201090rai  NaN  401326600          7          7             NaN   

      home_moneyline  spread_line  away_spread_odds  home_spread_odds  \
0                NaN         -4.0               NaN               NaN   
1                NaN         -3.0               NaN               NaN   
2                NaN         -6.0               NaN               NaN   
3                NaN  

In [1]:

pd.set_option('display.max_rows', 2000)
pd.set_option('display.max_columns', 400)

#Prediction data CSV
data = pd.read_csv('https://raw.githubusercontent.com/nflverse/nfldata/master/data/predictions.csv')

#Game results CSV
full_games_data = pd.read_csv('https://raw.githubusercontent.com/nflverse/nfldata/master/data/games.csv')

#Get median prediction for each game (not including Market)
#Some people also choose to exclude picks that are at 50% (not shown here)
pred_sum = data.loc[(data.screen_name!='Market')].groupby('game_id')[['prediction']].median()

#Market prediction for each game
pred_sum['market'] = data.loc[data.screen_name=='Market'].groupby('game_id')[['prediction']].sum()

#Full games data with results for this season
week_games_sum = full_games_data.loc[full_games_data.season==2021]

#Only keep select columns
week_games_sum = week_games_sum[['game_id','away_team','home_team','result']]

#If result column is > 0 then home team won
week_games_sum['winner'] = week_games_sum['result']
week_games_sum.loc[week_games_sum.result > 0, 'winner'] = week_games_sum.home_team
week_games_sum.loc[week_games_sum.result < 0, 'winner'] = week_games_sum.away_team

#Join to pred_sum
week_games_sum = week_games_sum.merge(pred_sum, left_on='game_id', right_index=True)

#Column for what the median user predicted winner was
week_games_sum['pred_winner'] = np.nan
week_games_sum.loc[week_games_sum.prediction > 50, 'pred_winner'] = week_games_sum.home_team
week_games_sum.loc[week_games_sum.prediction <= 50, 'pred_winner'] = week_games_sum.away_team

#Dataframe with every prediction as a row and the game result
points = data.merge(full_games_data[['game_id','home_team','away_team','week','season','result']],
                    left_on='game_id', right_on='game_id')

#Calculate points result for each prediction
points['pts'] = np.nan

#Formula for points
#M * (25 - (100 * (P/100 - R)^2))

#Neutral pick 50% gets 0 points
points.loc[points.prediction == 50, 'pts'] = 0

#Predicts Home Win, Actual Home Win
points.loc[(points.prediction > 50) & (points.result > 0), 'pts'] = 1 * (25 - (100 * (points.prediction/100 - 1)**2))

#Predicts Home Win, Actual Home Loss
points.loc[(points.prediction > 50) & (points.result < 0), 'pts'] = 1 * (25 - (100 * (points.prediction/100 - 0)**2))

#Create new column for away prediction
points['away_pred'] = np.nan

#Flip away predictions to the other side of 50 (i.e. 30 becomes 70)
points.loc[points.prediction < 50, 'away_pred'] = 100 - points.prediction 

#Predicts Away Win, Actual Away Win
points.loc[(points.away_pred > 50) & (points.result < 0), 'pts'] = 1 * (25 - (100 * (points.away_pred/100 - 1)**2))

#Predicts Away Win, Actual Away Loss
points.loc[(points.away_pred > 50) & (points.result > 0), 'pts'] = 1 * (25 - (100 * (points.away_pred/100 - 0)**2))

#Get each user's points
points.groupby('screen_name')[['pts']].sum().sort_values('pts')

#Get each game's aggregate points won/lost
points.groupby(['game_id','result','home_team','away_team'], as_index=False)[['pts']].sum()

game_id  result home_team away_team       pts
0    2021_01_ARI_TEN   -25.0       TEN       ARI  -5297.83
1     2021_01_BAL_LV     6.0        LV       BAL -14036.12
2     2021_01_CHI_LA    20.0        LA       CHI   8499.02
3     2021_01_CLE_KC     4.0        KC       CLE   6990.04
4     2021_01_DAL_TB     2.0        TB       DAL   8516.11
5    2021_01_DEN_NYG   -14.0       NYG       DEN   2343.13
6      2021_01_GB_NO    35.0        NO        GB -11907.86
7    2021_01_JAX_HOU    16.0       HOU       JAX  -7703.77
8    2021_01_LAC_WAS    -4.0       WAS       LAC    201.70
9     2021_01_MIA_NE    -1.0        NE       MIA  -3428.13
10   2021_01_MIN_CIN     3.0       CIN       MIN  -7786.17
11   2021_01_NYJ_CAR     5.0       CAR       NYJ   3153.98
12   2021_01_PHI_ATL   -26.0       ATL       PHI  -6672.98
13   2021_01_PIT_BUF    -7.0       BUF       PIT -15173.74
14   2021_01_SEA_IND   -12.0       IND       SEA   4600.65
15    2021_01_SF_DET    -8.0       DET        SF   9763.29
16    2021_02_ATL_TB    23.0        TB       ATL   8013.49
17   2021_02_BUF_MIA   -35.0       MIA       BUF   3297.55
18   2021_02_CIN_CHI     3.0       CHI       CIN  -1920.68
19   2021_02_DAL_LAC    -3.0       LAC       DAL  -1064.02
20   2021_02_DEN_JAX   -10.0       JAX       DEN   5144.21
21    2021_02_DET_GB    18.0        GB       DET   6846.17
22   2021_02_HOU_CLE    10.0       CLE       HOU   7711.90
23    2021_02_KC_BAL     1.0       BAL        KC  -8749.44
24    2021_02_LA_IND    -3.0       IND        LA   4661.95
25    2021_02_LV_PIT    -9.0       PIT        LV  -4710.56
26   2021_02_MIN_ARI     1.0       ARI       MIN   3907.65
27    2021_02_NE_NYJ   -19.0       NYJ        NE   5028.86
28    2021_02_NO_CAR    19.0       CAR        NO  -8305.05
29   2021_02_NYG_WAS     1.0       WAS       NYG   2334.82
30    2021_02_SF_PHI    -6.0       PHI        SF   2903.39
31   2021_02_TEN_SEA    -3.0       SEA       TEN  -8031.18
32   2021_03_ARI_JAX   -12.0       JAX       ARI   6393.06
33   2021_03_ATL_NYG    -3.0       NYG       ATL  -1764.49
34   2021_03_BAL_DET    -2.0       DET       BAL   6842.33
35   2021_03_CAR_HOU   -15.0       HOU       CAR   5182.80
36   2021_03_CHI_CLE    20.0       CLE       CHI   5224.46
37   2021_03_CIN_PIT   -14.0       PIT       CIN  -3690.25
38     2021_03_GB_SF    -2.0        SF        GB   -939.14
39   2021_03_IND_TEN     9.0       TEN       IND   3741.24
40    2021_03_LAC_KC    -6.0        KC       LAC  -7757.75
41    2021_03_MIA_LV     3.0        LV       MIA   4240.05
42     2021_03_NO_NE   -15.0        NE        NO  -1520.83
43   2021_03_NYJ_DEN    26.0       DEN       NYJ   6604.91
44   2021_03_PHI_DAL    20.0       DAL       PHI   3423.60
45   2021_03_SEA_MIN    13.0       MIN       SEA  -3684.11
46     2021_03_TB_LA    10.0        LA        TB  -2390.46
47   2021_03_WAS_BUF    22.0       BUF       WAS   5285.37
48    2021_04_ARI_LA   -17.0        LA       ARI  -6072.22
49   2021_04_BAL_DEN   -16.0       DEN       BAL   1474.82
50   2021_04_CAR_DAL     8.0       DAL       CAR   3075.59
51   2021_04_CLE_MIN    -7.0       MIN       CLE   1747.66
52   2021_04_DET_CHI    10.0       CHI       DET   -523.79
53   2021_04_HOU_BUF    40.0       BUF       HOU   8023.06
54   2021_04_IND_MIA   -10.0       MIA       IND  -1187.90
55   2021_04_JAX_CIN     3.0       CIN       JAX   4990.86
56    2021_04_KC_PHI   -12.0       PHI        KC   6352.95
57    2021_04_LV_LAC    14.0       LAC        LV   2455.40
58    2021_04_NYG_NO    -6.0        NO       NYG -10089.18
59    2021_04_PIT_GB    10.0        GB       PIT   5780.11
60    2021_04_SEA_SF    -7.0        SF       SEA   -862.19
61     2021_04_TB_NE    -2.0        NE        TB   5418.04
62   2021_04_TEN_NYJ     3.0       NYJ       TEN -11271.09
63   2021_04_WAS_ATL    -4.0       ATL       WAS    587.88
64    2021_05_BUF_KC   -18.0        KC       BUF  -2801.07
65    2021_05_CHI_LV   -11.0        LV       CHI  -5346.26
66   2021_05_CLE_LAC     5.0       LAC       CLE    354.09
6

#### 2021-11-07 use elo as predictions
- calc elo for each game
- output prediction probability
- ally games results, compare to elo predictions
- see how good prediction was?
    - what measure? error rate?
    - how to combine with probability prediction e.g 65% chance of Browns win
